### ISSUE TITLE
Reverse tethering being the tethered device the one with ADB access to the tethering device

### ISSUE THREAD

**Issue by rmnhg** [AlphaBs/termux-gnirehtet#2]
Hi.

I tried this tool and it works! Thank you for your work.

I got it working connecting the tethering phone to the tethered phone using a USB C male to USB A female adapter in the tethered phone, and a USB A male to micro USB male cable connected to the tethered phone.

However, I am not sure if I can do the same sharing if I use an adapter from micro USB male to USB A female and a USB A male to USB C male to the tethering phone.

Would it be possible? In my first attempt, I was not able to do it. I want to try the same being the tethered Android device a tablet with only USB A female ports, so I am not sure if it would work.

Is there any way I can get it working as I mention?

Thanks again!

**Comment by rmnhg** [comment]
I mean, I would like to be able to connect to the Internet from the device that can access the tethering device using ADB, like:
```
$tethered: adb devices:
* tethering
```

**Comment by AlphaBs** [comment]
so you mean,

`tethered - micro USB male - USB A female - USB A male - USB C female - tethering`?

I haven't tested it (I don't have usb male to female cable now), but it would work.

**Comment by rmnhg** [comment]
Exactly! I think I would need to use ADB to forward port 5037 (ADB port) from tethered to tethering device, so that the tethering device could connect to the tethered device through that port via ADB and run this script

**Comment by AlphaBs** [comment]
The tethered device must be a host device and the tethering device must be running an ADB server. it is impossible to do it the other way around since `gnirehtet` doesn't support it.

make sure that the tethered device is a host device if it doesn't work. (there will be an option like 'USB controlled by this device')

**Comment by rmnhg** [comment]
That's what I was thinking. The problem is that I wasn't able to perform that usb control switch (it does not work) and I didn't find another way to switch the USB controller device.
Therefore, my hope is that executing that adb forward I could perform a kind of reverse ADB connection and run this script.
Something like:
tethered:
adb forward tcp:5037 tcp:5038

tethering:
adb connect localhost:5038
./gnirehtet relay

But I'm looking for a termux alternative (perhaps creating an app) to run those ADB commands from the tethered device, since termux-api and termux-adb are not working, but other adb shell apps do

**Comment by AlphaBs** [comment]
I did it. connect two devices with USB cable. **the tethering device is host now.**

tethering device:
```
./gnirehtet relay # listening on 31416
```

tethered device:
turn on wireless debugging and pair with termux-adb
```
adb forward tcp:31416 tcp:31416 # forward gnirehtet relay server
adb connect localhost:40643 # connect itself
adb -s localhost:40643 reverse localabstract:gnirehtet tcp:31416
adb -s localhost:40643 shell am start -a com.genymobile.gnirehtet.START -n com.genymobile.gnirehtet/.GnirehtetActivity # start gnirehtet
```

**Comment by rmnhg** [comment]
Thank you for the steps @AlphaBs ! It is really nice to know it works.

My main concern with my setup is that my tethered device runs Android 10. Therefore, I can't enable wireless ADB from developer options.

Do you know if it would be possible to do the same without wireless ADB? Or to enable it without USB nor root permissions? Because that is becoming my main obstacle.

Another approach I have thought is to study the gnirehtet source to modify it to work with the 31416 port in the opposite direction, but first I need to understand better how it works.

**Comment by rmnhg** [comment]
Ok. I managed to simplify it without wireless ADB with these commands:

Tethering device:
`./gnirehtet relay # listening on 31416`

Tethered device;
```
adb forward localabstract:gnirehtet tcp:31416 # forward gnirehtet relay server
am start -a com.genymobile.gnirehtet.START -n com.genymobile.gnirehtet/.GnirehtetActivity # start gnirehtet
```

I'm running those commands in the tethered device **outside Termux and with root access**. I have to check if I can do the same without root permissions inside Termux or in another app. In my real tethered device, termux-adb is not working (it always says device offline) so I'm thinking about modifying the gnirehtet app to also implement ADB and establish this forwarding, but I'm not sure how easy it will be.

**Comment by AlphaBs** [comment]
on my device, termux-adb works well with `adb forward`. gnirehtet need to be modified to connect to tcp:31415. ([source code](https://github.com/Genymobile/gnirehtet/blob/1eb2e58bc91e268dba62561d11e3bebd9b100883/app/src/main/java/com/genymobile/gnirehtet/RelayTunnel.java#L49))
